### PR TITLE
feat: support building customized canvas entry

### PR DESCRIPTION
**Pull request by yy-wow**
English | [简体中文](https://github.com/opentiny/tiny-engine/blob/develop/.github/PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.zh-CN.md)

# PR

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our [Commit Message Guidelines](https://github.com/opentiny/tiny-engine/blob/develop/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Built its own designer, fully self-validated

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

### What is the current behavior?

Issue Number: N/A

### What is the new behavior?
1. 工具栏-保存增加beforeSave、saveMethod、saved钩子；
2. 工具栏-预览增加beforePreview、previewMethod、afterPreview钩子；
3. 画布iframe支持配置src替换默认的srcdoc，支持自定义的画布页面构建入口；
4. 导出获取组件condition值的方法

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Release Notes

- **New Features**
  - Enhanced iframe rendering logic in the Canvas components to support conditional sources.
  - Introduced new props for better handling of canvas sources and visibility states in various components.
  - Added new methods for managing conditions and improved state management capabilities.

- **Improvements**
  - Asynchronous handling added to several functions, improving the flow of operations during save and preview actions.
  - Added pre-save and post-save hooks for greater flexibility in handling method execution.
  - Streamlined control flow for save operations, enhancing user experience during loading states.
  - Improved interactivity and maintainability of components through method restructuring.

- **Bug Fixes**
  - Resolved issues related to method execution and state management within the components.

- **Configuration Changes**
  - Updated project to use ECMAScript modules, enhancing module import and export syntax.
  - Added `"type": "module"` property in `package.json` for modern JavaScript standards.

### CODE REVIEW

File: packages/toolbars/save/src/js/index.js [hunk 1]
```diff
@@ -95,6 +97,25 @@ export const openCommon = async () => {
     return
   }
 
+  const { beforeSave, saveMethod, saved } = getOptions(meta.id)
```
Comment (chilingling) [inline packages/toolbars/save/src/js/index.js:100]: 加一下文档，支持如何配置，适配的场景是什么？可以先简单加到 README 里面去

File: packages/toolbars/preview/src/Main.vue [hunk 2]
```diff
@@ -30,7 +31,26 @@ export default {
     const { isBlock, getCurrentPage, canvasApi } = useCanvas()
     const { getCurrentBlock } = useBlock()
 
-    const preview = () => {
+    const preview = async () => {
+      const { beforePreview, previewMethod, afterPreview } = getOptions(meta.id)
+      const printer = console
+
+      try {
+        if (typeof beforePreview === 'function') {
```
Comment (chilingling) [inline packages/toolbars/preview/src/Main.vue:39]: 这种 `beforexxx` 、 `xxx`、`afterxxx`  的生命周期感觉相对通用。
建议抽象成一个工具类。
另外，建议  `beforexxx`、`afterxxx` 支持配置数组，数组项里面是对应的一个个 `function`。

或者是通过事件订阅发布的形式进行触发，比如：
在  a 插件里面进行事件监听：

```javascript
subscribe('beforePreview', () => {
    // 自定义逻辑
})
```

File: packages/canvas/container/src/CanvasContainer.vue [hunk 3]
```diff
@@ -11,13 +11,22 @@
...
+        id="canvas"
+        ref="iframe"
+        :src="canvasSrc"
+        style="border: none; width: 100%; height: 100%"
+      ></iframe>
+      <iframe
+        v-else
+        id="canvas"
+        ref="iframe"
+        :srcdoc="canvasSrcDoc"
```
Comment (chilingling) [inline packages/canvas/container/src/CanvasContainer.vue]: 这里只有 `src` 与  `srcdoc` 两个属性的不同，建议通过动态属性名的方式进行支持

```jsx
<iframe
  :[attr]="value"
  // ... rest attr
>
</iframe>
```

Comment (yy-wow) [reply packages/canvas/container/src/CanvasContainer.vue]: 已修改

File: packages/toolbars/save/src/js/index.js (hunk 1 above)
Comment (yy-wow) [reply packages/toolbars/save/src/js/index.js:100]: 这具体到某个元应用的配置项了，放到README里是否不太合适

File: packages/canvas/render/src/context.js [hunk 4]
```diff
@@ -51,3 +51,5 @@ export const getContext = () => context
 export const setCondition = (id, visible = false) => {
   conditions[id] = visible
 }
+
+export const getCondition = () => conditions
```
Comment (hexqi) [inline packages/canvas/render/src/context.js]: getCondition应该与setCondition一致，接收id参数，查询/设置对应Id的condition

File: packages/canvas/render/src/runner.js [hunk 5]
```diff
@@ -49,7 +49,7 @@ const renderer = {
 const create = async (config) => {
   const { beforeAppCreate, appCreated } = config.lifeCycles || {}
   if (typeof beforeAppCreate === 'function') {
-    await beforeAppCreate({ Vue: { h }, canvasWin: window, api })
+    await beforeAppCreate({ api: renderer })
```
Comment (hexqi) [inline packages/canvas/render/src/runner.js:52]: vue h函数用不到了吗

File: packages/canvas/DesignCanvas/src/DesignCanvas.vue [hunk 6]
```diff
@@ -52,6 +55,13 @@ export default {
     const showMask = ref(true)
     const canvasRef = ref(null)
     let showModal = false // 弹窗标识
+    const { canvasSrc = '' } = getOptions(meta.id) || {}
```
Comment (hexqi) [inline packages/canvas/DesignCanvas/src/DesignCanvas.vue:58]: 模块的配置项需要有文档介绍

File: packages/canvas/render/src/runner.js (hunk 5 above)
Comment (yy-wow) [reply packages/canvas/render/src/runner.js:52]: 想到画布使用src替换srcdoc后，画布的vue依赖可以直接import使用，指向同个vue实例，就不需要传递出来了，已在本地验证

File: packages/toolbars/preview/src/Main.vue (hunk 2 above)
Comment (yy-wow) [reply packages/toolbars/preview/src/Main.vue:39]: 发布订阅机制待完善，当前适配到unibuy做构建验证

File: packages/toolbars/save/src/js/index.js (hunk 1 above)
Comment (yy-wow) [reply packages/toolbars/save/src/js/index.js:100]: 先补充到wiki，后面整理到统一文档

File: packages/canvas/render/src/context.js (hunk 4 above)
Comment (yy-wow) [reply packages/canvas/render/src/context.js]: 已修改getCondition实现，根据传入id获取condition。
另外增加导出getConditions api，攻获取conditions响应式数据

File: packages/canvas/DesignCanvas/src/DesignCanvas.vue (hunk 6 above)
Comment (yy-wow) [reply packages/canvas/DesignCanvas/src/DesignCanvas.vue:58]: 先写入到wiki文档，后面整理到统一文档

File: packages/canvas/index.js [hunk 7]
```diff
@@ -15,8 +15,9 @@ import { CanvasBreadcrumb } from './breadcrumb'
 import CanvasContainer from './container'
 import CanvasLayout from './layout'
 import DesignCanvas from './DesignCanvas'
+import { createRender } from './render'
 
-export { CanvasContainer, CanvasLayout, DesignCanvas }
+export { CanvasContainer, CanvasLayout, DesignCanvas, createRender }
```
Comment (rhlin) [inline packages/canvas/index.js]: 需要单独导出createRender 避免css污染

**Review by hexqi — COMMENTED** [review]
eslint报错、gpg签名问题需要处理

**Review by hexqi — COMMENTED** [review]
PR标题需要修改，不应该有unibuy字眼。==>  feat: ...(feature description)

**Pull request closed without merging**